File: libs/core-models/src/dataset/transaction.rs
//! Canonical Foundry-style dataset transaction primitives.
//!
//! This module is the single source of truth for the typed identifiers and
//! enums that describe dataset transactions across every OpenFoundry
//! service (data-asset-catalog, dataset-versioning, ingestion-replication,
//! pipeline-execution, ml-training, …). Services MUST consume these types
//! instead of redeclaring local strings or `Uuid` aliases so that the wire
//! format and Postgres encoding stay consistent.
//!
//! Wire/SQL encoding:
//! - [`TransactionId`] serializes as a UUID string.
//! - [`DatasetRid`] / [`BranchName`] serialize as plain strings.
//! - [`TransactionType`] / [`TransactionState`] serialize as lowercase
//!   strings (`"snapshot"`, `"committed"`, …) which matches the existing
//!   Postgres TEXT columns in `services/*/migrations/*.sql`.

use std::fmt;
use std::str::FromStr;

use serde::{Deserialize, Serialize};
use uuid::Uuid;

// ---------------------------------------------------------------------------
// Newtypes
// ---------------------------------------------------------------------------

/// Internal primary key of a dataset transaction (UUID v7 by convention).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(transparent)]
pub struct TransactionId(pub Uuid);

impl TransactionId {
    pub fn new() -> Self {
        Self(Uuid::now_v7())
    }

    pub fn into_inner(self) -> Uuid {
        self.0
    }
}

impl Default for TransactionId {
    fn default() -> Self {
        Self::new()
    }
}

impl fmt::Display for TransactionId {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        self.0.fmt(f)
    }
}

impl From<Uuid> for TransactionId {
    fn from(value: Uuid) -> Self {
        Self(value)
    }
}

impl From<TransactionId> for Uuid {
    fn from(value: TransactionId) -> Self {
        value.0
    }
}

/// Foundry-style dataset Resource Identifier.
///
/// Format: `ri.foundry.main.dataset.<uuid-v7>` (lowercase). The wrapper
/// performs structural validation but does not enforce the UUID variant
/// beyond parseability.
#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(transparent)]
pub struct DatasetRid(pub String);

/// Error returned when a string cannot be parsed as a [`DatasetRid`].
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct InvalidDatasetRid(pub String);

impl fmt::Display for InvalidDatasetRid {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "invalid dataset RID `{}` (expected `ri.foundry.main.dataset.<uuid>`)",
            self.0
        )
    }
}

impl std::error::Error for InvalidDatasetRid {}

impl DatasetRid {
    pub const PREFIX: &'static str = "ri.foundry.main.dataset.";

    /// Build a RID from a UUID, producing the canonical `ri.foundry…` form.
    pub fn from_uuid(uuid: Uuid) -> Self {
        Self(format!("{}{}", Self::PREFIX, uuid))
    }

    /// Mint a fresh RID backed by a new UUID v7.
    pub fn new() -> Self {
        Self::from_uuid(Uuid::now_v7())
    }

    /// Extract the UUID suffix, if the RID is well-formed.
    pub fn uuid(&self) -> Option<Uuid> {
        self.0
            .strip_prefix(Self::PREFIX)
            .and_then(|tail| Uuid::parse_str(tail).ok())
    }

    pub fn as_str(&self) -> &str {
        &self.0
    }
}

impl Default for DatasetRid {
    fn default() -> Self {
        Self::new()
    }
}

impl fmt::Display for DatasetRid {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(&self.0)
    }
}

impl FromStr for DatasetRid {
    type Err = InvalidDatasetRid;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let tail = s
            .strip_prefix(Self::PREFIX)
            .ok_or_else(|| InvalidDatasetRid(s.to_string()))?;
        Uuid::parse_str(tail).map_err(|_| InvalidDatasetRid(s.to_string()))?;
        Ok(Self(s.to_string()))
    }
}

/// Branch name within a dataset (Foundry default: `master`).
#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(transparent)]
pub struct BranchName(pub String);

/// Error returned when a branch name fails validation.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum InvalidBranchName {
    Empty,
    TooLong(usize),
    InvalidChar(char),
}

impl fmt::Display for InvalidBranchName {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::Empty => write!(f, "branch name must not be empty"),
            Self::TooLong(n) => write!(f, "branch name too long ({} chars, max 200)", n),
            Self::InvalidChar(c) => write!(
                f,
                "branch name contains invalid character `{}` (allowed: a-z A-Z 0-9 - _ / .)",
                c
            ),
        }
    }
}

impl std::error::Error for InvalidBranchName {}

impl BranchName {
    pub const MAX_LEN: usize = 200;

    /// The Foundry-default trunk branch name.
    pub fn master() -> Self {
        Self("master".to_string())
    }

    pub fn as_str(&self) -> &str {
        &self.0
    }

    pub fn into_string(self) -> String {
        self.0
    }
}

impl Default for BranchName {
    fn default() -> Self {
        Self::master()
    }
}

impl fmt::Display for BranchName {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(&self.0)
    }
}

impl FromStr for BranchName {
    type Err = InvalidBranchName;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        if s.is_empty() {
            return Err(InvalidBranchName::Empty);
        }
        if s.len() > Self::MAX_LEN {
            return Err(InvalidBranchName::TooLong(s.len()));
        }
        for ch in s.chars() {
            let ok = ch.is_ascii_alphanumeric() || matches!(ch, '-' | '_' | '/' | '.');
            if !ok {
                return Err(InvalidBranchName::InvalidChar(ch));
            }
        }
        Ok(Self(s.to_string()))
    }
}

// ---------------------------------------------------------------------------
// Enums
// ---------------------------------------------------------------------------

/// Foundry-style transaction operation taxonomy.
///
/// Mirrors the textual values currently stored in the `operation` column of
/// the `dataset_transactions` Postgres tables.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum TransactionType {
    /// Replaces the entire view contents (`SNAPSHOT` in Foundry).
    Snapshot,
    /// Adds new files/rows without rewriting prior data.
    Append,
    /// Logical row-level upsert (used by streaming + CDC).
    Update,
    /// Tombstone / removal of rows or files.
    Delete,
}

impl TransactionType {
    pub fn as_str(&self) -> &'static str {
        match self {
            Self::Snapshot => "snapshot",
            Self::Append => "append",
            Self::Update => "update",
            Self::Delete => "delete",
        }
    }
}

impl fmt::Display for TransactionType {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(self.as_str())
    }
}

impl FromStr for TransactionType {
    type Err = UnknownTransactionType;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s.to_ascii_lowercase().as_str() {
            "snapshot" => Ok(Self::Snapshot),
            "append" => Ok(Self::Append),
            "update" => Ok(Self::Update),
            "delete" => Ok(Self::Delete),
            other => Err(UnknownTransactionType(other.to_string())),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct UnknownTransactionType(pub String);

impl fmt::Display for UnknownTransactionType {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "unknown transaction type `{}`", self.0)
    }
}

impl std::error::Error for UnknownTransactionType {}

/// Lifecycle state of a dataset transaction.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum TransactionState {
    /// Started but not yet committed; writes are staged.
    Open,
    /// Successfully sealed; visible to readers.
    Committed,
    /// Cancelled or failed; staged writes are discarded.
    Aborted,
}

impl TransactionState {
    pub fn as_str(&self) -> &'static str {
        match self {
            Self::Open => "open",
            Self::Committed => "committed",
            Self::Aborted => "aborted",
        }
    }

    /// Whether this is a terminal state (no further transitions allowed).
    pub fn is_terminal(&self) -> bool {
        matches!(self, Self::Committed | Self::Aborted)
    }
}

impl fmt::Display for TransactionState {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(self.as_str())
    }
}

impl FromStr for TransactionState {
    type Err = UnknownTransactionState;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s.to_ascii_lowercase().as_str() {
            "open" => Ok(Self::Open),
            "committed" => Ok(Self::Committed),
            "aborted" => Ok(Self::Aborted),
            other => Err(UnknownTransactionState(other.to_string())),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct UnknownTransactionState(pub String);

impl fmt::Display for UnknownTransactionState {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "unknown transaction state `{}`", self.0)
    }
}

impl std::error::Error for UnknownTransactionState {}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn dataset_rid_roundtrip() {
        let uuid = Uuid::now_v7();
        let rid = DatasetRid::from_uuid(uuid);
        assert!(rid.as_str().starts_with(DatasetRid::PREFIX));
        assert_eq!(rid.uuid(), Some(uuid));
        let parsed: DatasetRid = rid.as_str().parse().expect("valid rid");
        assert_eq!(parsed, rid);
    }

    #[test]
    fn dataset_rid_rejects_bad_input() {
        assert!("ri.foundry.main.folder.123".parse::<DatasetRid>().is_err());
        assert!(
            "ri.foundry.main.dataset.not-a-uuid"
                .parse::<DatasetRid>()
                .is_err()
        );
    }

    #[test]
    fn branch_name_validation() {
        assert!("master".parse::<BranchName>().is_ok());
        assert!("feature/streaming-v2".parse::<BranchName>().is_ok());
        assert!("".parse::<BranchName>().is_err());
        assert!("bad name".parse::<BranchName>().is_err());
    }

    #[test]
    fn transaction_type_serde() {
        let json = serde_json::to_string(&TransactionType::Snapshot).unwrap();
        assert_eq!(json, "\"snapshot\"");
        let parsed: TransactionType = serde_json::from_str("\"append\"").unwrap();
        assert_eq!(parsed, TransactionType::Append);
        assert_eq!(
            "delete".parse::<TransactionType>().unwrap(),
            TransactionType::Delete
        );
    }

    #[test]
    fn transaction_state_terminal() {
        assert!(!TransactionState::Open.is_terminal());
        assert!(TransactionState::Committed.is_terminal());
        assert!(TransactionState::Aborted.is_terminal());
    }
}

File: libs/core-models/dataset/transaction.go
// Package dataset hosts the canonical Foundry-style dataset primitives:
// dataset RIDs, branch names and transaction state.
//
// Wire format mirrors the Rust source of truth verbatim:
//   - TransactionID: bare UUID string
//   - DatasetRID:    "ri.foundry.main.dataset.<uuid>" string
//   - BranchName:    plain string
//   - TransactionType / State: lowercase tokens
package dataset

import (
	"encoding/json"
	"errors"
	"fmt"
	"strings"

	"github.com/google/uuid"

	"github.com/openfoundry/openfoundry-go/libs/core-models/ids"
	"github.com/openfoundry/openfoundry-go/libs/core-models/rid"
)

// ---------------------------------------------------------------------------
// TransactionID
// ---------------------------------------------------------------------------

// TransactionID is the internal primary key of a dataset transaction (UUID v7).
type TransactionID struct {
	uuid.UUID
}

// NewTransactionID mints a fresh time-ordered transaction id.
func NewTransactionID() TransactionID { return TransactionID{UUID: ids.New()} }

func (t TransactionID) MarshalJSON() ([]byte, error) {
	return json.Marshal(t.UUID.String())
}

func (t *TransactionID) UnmarshalJSON(data []byte) error {
	var s string
	if err := json.Unmarshal(data, &s); err != nil {
		return err
	}
	parsed, err := uuid.Parse(s)
	if err != nil {
		return fmt.Errorf("transaction id: %w", err)
	}
	t.UUID = parsed
	return nil
}

// ---------------------------------------------------------------------------
// DatasetRID
// ---------------------------------------------------------------------------

// DatasetRIDPrefix is the canonical prefix every dataset RID carries.
const DatasetRIDPrefix = "ri.foundry.main.dataset."

// DatasetRID is the Foundry-style resource identifier for a dataset.
type DatasetRID string

// NewDatasetRID mints a brand-new RID backed by a v7 UUID.
func NewDatasetRID() DatasetRID {
	return DatasetRID(rid.MustMintUUIDV7("foundry", rid.DefaultInstance, "dataset").String())
}

// DatasetRIDFromUUID builds the canonical "ri.foundry.main.dataset.<uuid>" form.
func DatasetRIDFromUUID(id uuid.UUID) DatasetRID {
	return DatasetRID(rid.MustNewUUID("foundry", rid.DefaultInstance, "dataset", id).String())
}

// String returns the RID in its canonical form.
func (r DatasetRID) String() string { return string(r) }

// UUID extracts the UUID suffix; returns false if the RID is malformed.
func (r DatasetRID) UUID() (uuid.UUID, bool) {
	parsed, err := rid.ParseUUID(string(r))
	if err != nil ||
		parsed.Service != "foundry" ||
		parsed.Instance != rid.DefaultInstance ||
		parsed.ResourceType != "dataset" {
		return uuid.UUID{}, false
	}
	id, err := uuid.Parse(parsed.Locator)
	if err != nil {
		return uuid.UUID{}, false
	}
	return id, true
}

// ParseDatasetRID validates and returns a DatasetRID.
func ParseDatasetRID(s string) (DatasetRID, error) {
	rid := DatasetRID(s)
	if _, ok := rid.UUID(); !ok {
		return "", &InvalidDatasetRIDError{Input: s}
	}
	return rid, nil
}

// InvalidDatasetRIDError is returned when ParseDatasetRID fails.
type InvalidDatasetRIDError struct{ Input string }

func (e *InvalidDatasetRIDError) Error() string {
	return fmt.Sprintf(
		"invalid dataset RID %q (expected %s<uuid>)",
		e.Input, DatasetRIDPrefix,
	)
}

// ---------------------------------------------------------------------------
// BranchName
// ---------------------------------------------------------------------------

// BranchName is a validated dataset branch name.
type BranchName string

const (
	// BranchMaster is the Foundry-default trunk branch.
	BranchMaster BranchName = "master"
	// BranchNameMaxLen is the upper bound enforced by ParseBranchName.
	BranchNameMaxLen = 200
)

// InvalidBranchNameError describes why a branch name failed validation.
type InvalidBranchNameError struct {
	Input  string
	Reason string
}

func (e *InvalidBranchNameError) Error() string {
	return fmt.Sprintf("invalid branch name %q: %s", e.Input, e.Reason)
}

// ParseBranchName validates the branch name and returns a typed value.
func ParseBranchName(s string) (BranchName, error) {
	switch {
	case len(s) == 0:
		return "", &InvalidBranchNameError{Input: s, Reason: "branch name must not be empty"}
	case len(s) > BranchNameMaxLen:
		return "", &InvalidBranchNameError{
			Input:  s,
			Reason: fmt.Sprintf("branch name too long (%d chars, max %d)", len(s), BranchNameMaxLen),
		}
	}
	for _, ch := range s {
		ok := (ch >= 'a' && ch <= 'z') ||
			(ch >= 'A' && ch <= 'Z') ||
			(ch >= '0' && ch <= '9') ||
			ch == '-' || ch == '_' || ch == '/' || ch == '.'
		if !ok {
			return "", &InvalidBranchNameError{
				Input:  s,
				Reason: fmt.Sprintf("invalid character %q (allowed: a-z A-Z 0-9 - _ / .)", ch),
			}
		}
	}
	return BranchName(s), nil
}

// ---------------------------------------------------------------------------
// TransactionType
// ---------------------------------------------------------------------------

// TransactionType is the operation taxonomy stored in `dataset_transactions.operation`.
type TransactionType string

const (
	TxSnapshot TransactionType = "snapshot"
	TxAppend   TransactionType = "append"
	TxUpdate   TransactionType = "update"
	TxDelete   TransactionType = "delete"
)

// ErrUnknownTransactionType is returned when a string token can't be parsed.
var ErrUnknownTransactionType = errors.New("unknown transaction type")

// ParseTransactionType is case-insensitive and matches the Rust FromStr impl.
func ParseTransactionType(s string) (TransactionType, error) {
	switch strings.ToLower(s) {
	case "snapshot":
		return TxSnapshot, nil
	case "append":
		return TxAppend, nil
	case "update":
		return TxUpdate, nil
	case "delete":
		return TxDelete, nil
	default:
		return "", fmt.Errorf("%w: %q", ErrUnknownTransactionType, s)
	}
}

// ---------------------------------------------------------------------------
// TransactionState
// ---------------------------------------------------------------------------

// TransactionState is the lifecycle state of a dataset transaction.
type TransactionState string

const (
	TxOpen      TransactionState = "open"
	TxCommitted TransactionState = "committed"
	TxAborted   TransactionState = "aborted"
)

// ErrUnknownTransactionState is returned when a string token can't be parsed.
var ErrUnknownTransactionState = errors.New("unknown transaction state")

// ParseTransactionState is case-insensitive and matches the Rust FromStr impl.
func ParseTransactionState(s string) (TransactionState, error) {
	switch strings.ToLower(s) {
	case "open":
		return TxOpen, nil
	case "committed":
		return TxCommitted, nil
	case "aborted":
		return TxAborted, nil
	default:
		return "", fmt.Errorf("%w: %q", ErrUnknownTransactionState, s)
	}
}

// IsTerminal reports whether the state forbids further transitions.
func (s TransactionState) IsTerminal() bool {
	return s == TxCommitted || s == TxAborted
}
